.github/workflows/tox-pytest.yml

```diff
@@ -0,0 +1,93 @@
+name: tox-pytest
+
+on: [ push, pull_request ]
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: ["3.8", "3.9", "3.10"]
+      fail-fast: false
+    defaults:
+      run:
+        shell: bash -l {0}
+
+    steps:
+    - uses: actions/checkout@v2.4.0
+      with:
+        fetch-depth: 2
+
+    - name: Set up conda environment for testing
+      uses: conda-incubator/setup-miniconda@v2.1.1
+      with:
+        mamba-version: "*"
+        channels: conda-forge,defaults
+        channel-priority: true
+        python-version: ${{ matrix.python-version }}
+        activate-environment: pudl-rmi
+        environment-file: environment.yml
+
+    - name: Log the conda environment
+      run: |
+        conda info
+        conda list
+        conda config --show-sources
+        conda config --show
+        printenv | sort
+
+    - name: Cache PUDL DB
+      uses: actions/cache@v2
+      with:
+        path: ~/pudl-work/sqlite/pudl.sqlite
+        # The last time we deployed Datasette. Update if we redeploy.
+        key: "2021-11-12"
+
+    - name: Download PUDL DB and log pre-test PUDL workspace contents
+      run: |
+        mkdir -p ~/pudl-work/sqlite/
+        if [ ! -f ~/pudl-work/sqlite/pudl.sqlite ]
+        then curl -o ~/pudl-work/sqlite/pudl.sqlite https://data.catalyst.coop/pudl.db
+        fi
+        find ~/pudl-work/
+
+    - name: Log SQLite3 version
+      run: |
+        conda run -n pudl-rmi which sqlite3
+        conda run -n pudl-rmi sqlite3 --version
+
+    - name: Run pytest using Tox
+      env:
+        API_KEY_EIA: ${{ secrets.API_KEY_EIA }}
+      run: |
+        conda run -n pudl-rmi tox
+
+    - name: Log post-test PUDL workspace contents
+      run: find ~/pudl-work/
+
+    - name: Upload test coverage report to CodeCov
+      uses: codecov/codecov-action@v2.1.0
+      with:
+        verbose: true
+        files: ./coverage.xml
+
+    - name: Inform the Codemonkeys
+      uses: 8398a7/action-slack@v3
+      with:
+        status: custom
+        fields: workflow,job,commit,repo,ref,author,took
+        custom_payload: |
+          {
+            username: 'action-slack',
+            icon_emoji: ':octocat:',
+            attachments: [{
+              color: '${{ job.status }}' === 'success' ? 'good' : '${{ job.status }}' === 'failure' ? 'danger' : 'warning',
+              text: `${process.env.AS_WORKFLOW}\n${process.env.AS_JOB} (${process.env.AS_COMMIT}) of ${process.env.AS_REPO}@${process.env.AS_REF} by ${process.env.AS_AUTHOR} ${{ job.status }} in ${process.env.AS_TOOK}`,
+            }]
+          }
```

Comment:
What do you think about creating a separate slack channel for each project? That way non RMI folks won't be distracted by RMI build notifications. Also, adding CI to our private repos should post to private slack channels.

Reply:
A separate GitHub CI notification channel for each project?  Yeah that probably makes sense. So many channels. I always find setting up the webooks to be a hassle and I think we'd need a separate one for every channel.

Reply:
I've added a repository specific `SLACK_WEBHOOK_URL` secret which will notify the `#rmi-builds` channel in our Slack... assuming this is all setup right.

Though I feel like I've noticed lately that we get notified of **passing** builds, but not **failing** builds, which seems bad and wrong.

Reply:
This is great thank you. I haven't noticed any missing failing builds in #codemonkeys.